### ISSUE TITLE
ABI changed, but version of shared library did not?

### ISSUE THREAD

**Issue by dankegel** [ddccontrol/ddccontrol#58]
I tried the following on Ubuntu 18.04 (bear with me):

$ sudo apt remove ddccontrol libddccontrol0 ddccontrol-db || true
$ git clone git@github.com:ddccontrol/ddccontrol-db.git
$ cd ddcontrol-db
$ ./autogen.sh
$ ./configure --prefix=/usr/
$ sudo make install
$ cd ..
$ git clone git@github.com:ddccontrol/ddccontrol.git
$ cd ddccontrol
$ ./autogen.sh
$ ./configure --prefix=/usr/ --sysconfdir=/etc --libexecdir=/usr/lib
$ make
$ src/ddccontrol/ddccontrol -p

This worked at least well enough to not crash.  However, then doing

$ sudo apt install libddccontrol0 
$ src/ddccontrol/ddccontrol -p

crashed (and locked the screen for a while).  This seems related:

$ ldd ./src/ddccontrol/.libs/ddccontrol
	linux-vdso.so.1 (0x00007ffc725b1000)
	libddccontrol.so.0 => /usr/lib/x86_64-linux-gnu/libddccontrol.so.0 (0x00007fe5eb0c2000)
	libddccontrol_dbus_client.so.0 => /usr/lib/libddccontrol_dbus_client.so.0 (0x00007fe5eaeb9000)

It seems ddccontrol is finding the distro's version of libddccontrol.so.  This is especially bad because
the ABI of libddccontrol.so changed incompatibly (struct monitor changed size and gained a vtable) as of https://github.com/ddccontrol/ddccontrol/commit/137b52f13f484fd3562379c40b45de3b1ac4dcaa

Since ddccontrol from git is not released, that's fine, it's not at all unexpected.

But before you do your next release, please bump the .0 on the end of that shared library
filename to .1 to allow apps that depend on your old version to keep running!
See e.g. the bottom of https://autotools.io/libtool/version.html

(I know, ABI stability is a pain, but it's really important.  You have to arrange for old and new
versions to coexist on same machine if you want rollouts that don't drive users screaming to
your issue tracker.)

**Comment by kravemir** [comment]
Thanks for filling the issue. I'll take a better look at it later, definitely before releasing.

Well, library should end with `.0.4.4`, or the next version. As it's still `0.x.x` version, it means it can be unstable.

Or, do you want to bump the version to the `.1.x.x`?

But, I don't like to break any ABI/API. That's one of reasons, why I'm considering to fully encapsulate structures, expose only pointers, and provide accessor methods. This would make one ABI/API breakage,... But, then it should be much more stable.

**Comment by kravemir** [comment]
> https://autotools.io/libtool/version.html

Took a quick look,.. Seems to be something different, than software version. Or?

**Comment by dankegel** [comment]
Since the library has shipped and is installed on user systems,
it's by definition stable (remember Linus's rule: don't break user systems).
So if you change the ABI significantly (and this was a very significant change),
you probably need to bump the major version if you want to avoid user complaints.

I've been staring at libtool doc for 20 years and still don't know the right thing to do :-)
The "ABI Versioning" section of https://www.akkadia.org/drepper/dsohowto.pdf may provide some guidance.
If in doubt, use the coarsest mechanism (just change major in filename, leave everything else zero), maybe.

Good luck!

**Comment by kravemir** [comment]
Now I'm reading, that library version shouldn't match package/software version. From https://autotools.io/libtool/version.html:

> Another common mistake is to match the value of -version-info with the package version or vice-versa...

And from https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html:

> Never try to set the interface numbers so that they correspond to the release number of your package. This is an abuse that only fosters misunderstanding of the purpose of library versions...

So, regarding library versioning, following applies, again from https://autotools.io/libtool/version.html:

> The rules of thumb, when dealing with these values are:
> 
> - Always increase the revision value.
> - Increase the current value whenever an interface has been added, removed or changed.
> - Increase the age value only if the changes made to the ABI are backward compatible.

Or from https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html:

> Here are a set of rules to help you update your library version information:
> 
> 1. Start with version information of ‘0:0:0’ for each libtool library.
> 2. Update the version information only immediately before a public release of your software. More frequent updates are unnecessary, and only guarantee that the current interface number gets larger faster.
> 3. If the library source code has changed at all since the last update, then increment revision (‘c:r:a’ becomes ‘c:r+1:a’).
> 4. If any interfaces have been added, removed, or changed since the last update, increment current, and set revision to 0.
> 5. If any interfaces have been added since the last public release, then increment age.
> 6. If any interfaces have been removed or changed since the last public release, then set age to 0.

> The following explanation may help to understand the above rules a bit better: consider that there are three possible kinds of reactions from users of your library to changes in a shared library:
> 
> 1. Programs using the previous version may use the new version as drop-in replacement, and programs using the new version can also work with the previous one. In other words, no recompiling nor relinking is needed. In this case, bump revision only, don’t touch current nor age.
> 2. Programs using the previous version may use the new version as drop-in replacement, but programs using the new version may use APIs not present in the previous one. In other words, a program linking against the new version may fail with “unresolved symbols” if linking against the old version at runtime: set revision to 0, bump current and age.
> 3. Programs may need to be changed, recompiled, and relinked in order to use the new version. Bump current, set revision and age to 0.

**Comment by kravemir** [comment]
> Since the library has shipped and is installed on user systems,
> it's by definition stable (remember Linus's rule: don't break user systems).

Agree. I'd rather make user/maintainer be informed, that application/package needs a newer library version, than caused undefined behaviour, which can be sometimes hard to notice, because it just corrupts data,.. and then it's harder to trace issue.

Thank you for informing about ABI change issues by filling this issue.

**Comment by dankegel** [comment]
Sounds like you're all set.  Only thing to do at the moment is perhaps to add a warning about possible conflict with distro version if installing from source.

**Comment by dankegel** [comment]
You could, if you wanted, go with something like the odd-even version numbering scheme common in linux.
When an ABI-breaking change is made, increment current and clear revision.  Then at release time, if there were any more abi-breaking changes, increment it again.  That way you don't increment on every little break, just on the first and last break during development of a release.
